lib/date_time/date_time.c

```diff
@@ -123,27 +190,137 @@ static int time_modem_get(void)
+	/* 4 MSBs represent decimal that should be added as is. */
+	return ((value & 0xf0) >> 4) + ((value & 0x0f) * 10);
+}
+
+#if defined(CONFIG_DATE_TIME_AUTO)
+static void date_time_at_xtime_handler(const char *notif)
+{
+	struct tm date_time;
+	uint8_t time_buf[6];
+	size_t time_buf_len;
+	char *time_str_start;
+
+	if (notif == NULL) {
+		return;
+	}
+
+	/* %XTIME: [<local_time_zone>],[<universal_time>],[<daylight_saving_time>]"
+	 * <unversal_time> string in hex format, seven bytes long optional field for
+	 *                 universal time as specified in 3GPP TS 24.008 ch 10.5.3.9
+	 *                 and received from network.
+	 * Examples of modem time response:
+	 * %XTIME: "08","81109251714208","01"
+	 * %XTIME: ,"81109251714208",
+	 */
+	time_str_start = strchr(notif, ',');
+	if (time_str_start == NULL) {
+		LOG_ERR("%%XTIME notification doesn't contain ',': %s", notif);
+		return;
+	}
+	if (strlen(time_str_start) < 17) {
+		LOG_ERR("%%XTIME notification too short: %s", notif);
+		return;
+	}
+	if (*(time_str_start + 1) != '"') {
+		LOG_ERR("%%XTIME notification doesn't contain '\"' after ',': %s", notif);
+		return;
+	}
+
+	time_str_start += 2;
+	time_buf_len = hex2bin(time_str_start, 12, time_buf, sizeof(time_buf));
+
+	if (time_buf_len < sizeof(time_buf)) {
+		LOG_ERR("%%XTIME notification decoding failed (ret=%d): %s", time_buf_len, notif);
+	}
+
+	modem_valid_time = true;
+
+	date_time.tm_year = semioctet_to_dec(time_buf[0]);
+	date_time.tm_mon  = semioctet_to_dec(time_buf[1]);
+	date_time.tm_mday = semioctet_to_dec(time_buf[2]);
+	date_time.tm_hour = semioctet_to_dec(time_buf[3]);
+	date_time.tm_min  = semioctet_to_dec(time_buf[4]);
+	date_time.tm_sec  = semioctet_to_dec(time_buf[5]);
+
+	LOG_DBG("Time: %u/%u/%u,%u:%u:%u",
+		date_time.tm_year,
+		date_time.tm_mon,
+		date_time.tm_mday,
+		date_time.tm_hour,
+		date_time.tm_min,
+		date_time.tm_sec);
+
+	/* Relative to 1900, as per POSIX */
+	date_time.tm_year = date_time.tm_year + 2000 - 1900;
+	/* Range is 0-11, as per POSIX */
+	date_time.tm_mon = date_time.tm_mon - 1;
+
+	time_aux.date_time_utc = (int64_t)timeutil_timegm64(&date_time) * 1000;
+	time_aux.last_date_time_update = k_uptime_get();
+
+	date_time_store(time_aux.date_time_utc);
+
+	LOG_DBG("Time obtained from cellular network (XTIME notification)");
+}
+#endif /* defined(CONFIG_DATE_TIME_AUTO) */
+#endif /* defined(CONFIG_DATE_TIME_MODEM) */
+
+#if defined(CONFIG_LTE_LINK_CONTROL)
+
+void date_time_lte_ind_handler(const struct lte_lc_evt *const evt)
+{
+	switch (evt->type) {
+	case LTE_LC_EVT_NW_REG_STATUS:
+		LOG_DBG("LTE registration status: %d", evt->nw_reg_status);
+		if (evt->nw_reg_status == LTE_LC_NW_REG_REGISTERED_EMERGENCY ||
+		    evt->nw_reg_status == LTE_LC_NW_REG_REGISTERED_HOME ||
+		    evt->nw_reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING) {
+			/* TODO: Maybe use delayable work with for example 1s delay to make
+			 * it more probably that modem has received time.
+			 */
```

Comment:
Sounds like a good idea, it does take some time after the +CEREG notification before the time is valid and the work might be scheduled too fast without any delay.

Reply:
Yep, even without the delay, DNA and Elisa networks seem to provide time with the current implementation but to be more compliant globally, we'd probably want some delay. Whether it's 1s, 3s or 5s, I have no idea. Maybe not going beyond 5s.

Reply:
With Telia and DNA the delay between LTE attach complete and the EMM info message was 100 ms and with Elisa it was less than 40 ms. I propose a 1 second delay. I don't think we should delay it excessively based on the possibility that in some network it may take longer than that.